routing_common/transit_graph_data.cpp

```diff
@@ -0,0 +1,586 @@
+  edgeFlags.SetFlags(0);
+  edgeFlags.m_transfer = transfer;
+}
+
+void DeserializerFromJson::operator()(StopIdRanges & rs, char const * name)
+{
+  vector<StopId> stopIds;
+  (*this)(stopIds, name);
+  rs = StopIdRanges({stopIds});
+}
+
+// GraphData --------------------------------------------------------------------------------------
+void GraphData::DeserializeFromJson(my::Json const & root, OsmIdToFeatureIdsMap const & mapping)
+{
+  DeserializerFromJson deserializer(root.get(), mapping);
+  Visit(deserializer);
+
+  // Removes equivalent edges from |m_edges|. If there are several equivalent edges only
+  // the most lightweight edge is left.
+  // Note. It's possible that two stops are connected with the same line several times
+  // in the same direction. It happens in Oslo metro (T-banen):
+  // https://en.wikipedia.org/wiki/Oslo_Metro#/media/File:Oslo_Metro_Map.svg branch 5.
+  my::SortUnique(m_edges,
+                 [](Edge const & e1, Edge const & e2) {
+                   if (e1 != e2)
+                     return e1 < e2;
+                   return e1.GetWeight() < e2.GetWeight();
+                 },
+                 [](Edge const & e1, Edge const & e2) { return e1 == e2; });
+}
+
+void GraphData::Serialize(Writer & writer)
+{
+  auto const startOffset = writer.Pos();
+  Serializer<Writer> serializer(writer);
+  FixedSizeSerializer<Writer> numberSerializer(writer);
+  m_header.Reset();
+  numberSerializer(m_header);
+  m_header.m_stopsOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_stops);
+  m_header.m_gatesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_gates);
+  m_header.m_edgesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_edges);
+  m_header.m_transfersOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_transfers);
+  m_header.m_linesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_lines);
+  m_header.m_shapesOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_shapes);
+  m_header.m_networksOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  serializer(m_networks);
+  m_header.m_endOffset = base::checked_cast<uint32_t>(writer.Pos() - startOffset);
+
+  // Rewriting header info.
+  CHECK(m_header.IsValid(), (m_header));
+  auto const endOffset = writer.Pos();
+  writer.Seek(startOffset);
+  numberSerializer(m_header);
+  writer.Seek(endOffset);
+
+  LOG(LINFO, (TRANSIT_FILE_TAG, "section is ready. Header:", m_header));
+}
+
+void GraphData::DeserializeAll(Reader & reader)
+{
+  NonOwningReaderSource src(reader);
+
+  ReadHeader(src);
+  ReadStops(src);
+  ReadGates(src);
+  ReadEdges(src);
+  ReadTransfers(src);
+  ReadLines(src);
+  ReadShapes(src);
+  ReadNetworks(src);
+}
+
+void GraphData::DeserializeForRouting(Reader & reader)
+{
+  NonOwningReaderSource src(reader);
```

Comment:
Code duplication here - initialization of reader source, deserialization of header.

Reply:
And deserialization stops. We can refactor it in several ways:
1. to make a method which initialize reader source, deserialize header and stops. And use it.
2. to make a method which initialize reader source, deserialize header. And use it.
3. to rewrite the code a little using enum and one method instead of DeserializeAll(), DeserializeForRouting(), DeserializeForRendering(), DeserializeForCrossMwm()

What do you think about that?

Reply:
Actually we may leave this as is - but duplication is really bad. You may use continuations, for example:
```c++
template <typename Fn>
void DeserializeWith(Reader & reader, Fn && fn)
{
  NonOwningReaderSource src(reader);
  ReadHeader(src);
  fn(src);
}
```
and implement three methods, say:
```c++
void DeserializeForRouting(Reader & reader)
{
  DeserializeWith(reader, [&](NonOwningReaderSource & src)
  {
    DeserializeShapes(src);
    ...
  });
}
```
but it doesn't seem to me as a great improvement. WDYT?

Reply:
Taking into account that DoSomethingWith() method (continuations) is widely spread in our project I'm for this variant. So I'll rewrite it as you proposed.